src/main.rs

```diff
@@ -54,300 +43,29 @@ fn parse_interval(
-
-        let mods_xor1 = mods_raw & 0xFFFFFFFF;
-        let mods_xor2 = mods_raw >> 32;
-
-        values.mods = (mods_xor1 ^ mods_xor2) as u32;
-
-        // Calculate pp
-        values.current_pp = values.get_current_pp();
-        values.fc_pp = values.get_fc_pp();
-
-        values.prev_passed_objects = passed_objects;
-        
-        values.grade = values.get_current_grade();
-        values.current_bpm = values.get_current_bpm();
-        values.kiai_now = values.get_kiai();
-
-        // Placing at the very end cuz we should
-        // keep up with current_bpm & unstable rate
-        // updates
-        values.adjust_bpm();
-    }
-
-    Ok(())
-}
-fn handle_clients(
-    values: &Values,
-    clients: &mut HashMap<usize, WebSocketStream<Async<TcpStream>>>
-) {
-    let _span = span!("handle clients");
-    clients.retain(|_client_id, websocket| {
-        smol::block_on(async {
-            let _span = span!("send message to clients");
-
-            let next_future = websocket.next();
-            let msg_future =
-                smol::future::poll_once(next_future);
-
-            #[allow(clippy::collapsible_match)]
-            let msg = match msg_future.await {
-                Some(v) => {
-                    match v {
-                        Some(Ok(v)) => Some(v),
-                        Some(Err(_)) => return false,
-                        None => None,
-                    }
-                },
-                None => None,
-            };
-
-
-            if let Some(tungstenite::Message::Close(_)) = msg {
-                return false;
-            };
-
-            let _ = websocket.send(
-                Message::Text(
-                    serde_json::to_string(&values)
-                    .unwrap()
-                    ) // No way serialization gonna fail so
-                      // using unwrap
-                ).await;
-
-            true
-        })
-    });
-}
 
 fn main() -> Result<()> {
     let _client = tracy_client::Client::start();
 
     let args = Args::parse();
-    let mut values = Values::default();
-
-    let (tx, rx) = bounded::<WebSocketStream<Async<TcpStream>>>(20);
-
-    std::thread::spawn(move || server_thread(tx.clone()));
-
-    let mut client_id = 0;
-    let mut clients: HashMap<usize, WebSocketStream<Async<TcpStream>>> = 
-        HashMap::new();
-
-    let mut static_static_addresses = StaticAddresses::default();
+    let output_values = Arc::new(Mutex::new(OutputValues::default()));
+    let inner_values = InnerValues::default();
+
+    let mut state = State {
+        addresses: StaticAddresses::default(),
+        clients: Clients::default(),
+        ivalues: inner_values,
+        values: output_values,
+    };
     
-    // TODO ugly nesting mess
+    // Spawning Hyper server
+    let server_clients = state.clients.clone();
+    std::thread::spawn(move || server_thread(server_clients));
+
     'init_loop: loop {
+        let mut values = state.values.lock().unwrap();
```

Comment:
`MutexGuard` should always be hold in as small of a scope as possible so this line should happen after the next `match`.

Reply:
thanks, placed lock acquiring after next `match`